### DISCUSSION TITLE
Control panel plugin

### DISCUSSION THREAD

**Discussion by dresco** [grblHAL/core#82] [Show and tell]
I've been fooling around with the idea of a hardware control panel for my grblHAL router for a while, inspired originally by my own clumsiness (have had a few 'whoops' moments when trying to look at both the machine and the display and mouse click on the right controls).

Is definitely not trying to replace a software sender, but just to add another layer of physical controls to make things a bit more tactile & intuitive. Buttons for various commands, encoders for jogging and overrides, along with an LCD display to view the current position & state etc...

To this end, I've got the beginnings of a plugin that talks to my PoC panel hardware over RS485 modbus. Still very much a work in progress, but it can be found at https://github.com/dresco/Plugin_panel

This will likely become more interesting when there is an example panel to try it out with, which is what I'll be looking at next. Will all be open source, and I'm hoping to produce a reference design that could be ordered from somewhere like seeed or tindie to make it easier for anyone interested to get started with.

**Comment by terjeio** [comment]
Great. FYI I now have added information about [how to add a third-party plugin](https://github.com/grblHAL/plugins#i-have-written-a-plugin-and-i-want-to-make-it-available-to-grblhal-users) to grblHAL.
Should I add a link to yours in the plugin repo readme?

**Reply by dresco** [reply]
> Should I add a link to yours in the plugin repo readme?
Yes please, thanks!

**Comment by dresco** [comment]
Have recently been making some progress with a reference control panel implementation, and have updated the project repositories to reflect the current state.. The new top level repository can be found [here](https://github.com/dresco/grblPANEL) 

I've been hand assembling my prototypes, which doesn't really scale. But perhaps if there is sufficient interest, I could see about getting boards assembled offshore, and offer a kit of the important parts?

Regardless, would be interested in any feedback, cheers!

**Reply by andrewmarles** [reply]
Looks amazing.  Very nice work.

**Reply by cbaugher** [reply]
@dresco 

This is extremely cool and I want one!  Are there PCBs available for purchase?  The mechanical parts and assembly I can handle, and I could even have PCBs made if need be.  But it would be great to get a jump start if I could skip that step.

I'm doing a controls update on an old Denford Novaturn lathe.  I'm planning to use a pi4 for gcode sending and was hoping to be able to have physical controls for jogging and such.  This fits the need almost perfectly!

C|

**Comment by dresco** [comment]
> This is extremely cool and I want one! Are there PCBs available for purchase? The mechanical parts and assembly I can handle, and I could even have PCBs made if need be. But it would be great to get a jump start if I could skip that step.
> 

Hi @cbaugher, glad you like it!

I have spare front panel, keypad, and encoder PCBs, plus a couple of displays. 

The sticking point is probably the main PCB, I've hand assembled my first few prototypes - which hasn't been tons of fun ;) A few of the parts weren't available as standard JLC/LCSC parts, and it needs some double sided SMT assembly (just because of the way it's mounted behind the screen). 

I believe both of those points can be addressed (using JLC parts consignment and paying extra for double sided assembly), but I've not looked into it yet. Guess I should work out the costs, so I know how much interest I'd need to make it feasible.

> I'm doing a controls update on an old Denford Novaturn lathe. I'm planning to use a pi4 for gcode sending and was hoping to be able to have physical controls for jogging and such. This fits the need almost perfectly!
> 

Nice, where abouts are you based, am guessing also UK with a Denford?

**Comment by cbaugher** [comment]
> Hi @cbaugher, glad you like it!
> 
> I have spare front panel, keypad, and encoder PCBs, plus a couple of displays.
> 
> The sticking point is probably the main PCB, I've hand assembled my first few prototypes - which hasn't been tons of fun ;) A few of the parts weren't available as standard JLC/LCSC parts, and it needs some double sided SMT assembly (just because of the way it's mounted behind the screen).
> 
> I believe both of those points can be addressed (using JLC parts consignment and paying extra for double sided assembly), but I've not looked into it yet. Guess I should work out the costs, so I know how much interest I'd need to make it feasible.
> 

It would be great to have the SMT parts pre-assembled, but I can do some amount of SMT assembly so that wouldn't be a problem for me.  As long as all the parts are currently available in the USA.


> 
> Nice, where abouts are you based, am guessing also UK with a Denford?

In the USA actually, near Chicago.  A few Denfords did make it over here, mostly for tech schools I think.  It seems to be a well made machine, though I've only had it in my possession for a few days.  I haven't even powered it up yet, and the way things are going I'll probably end up tearing into it right away.  LOL

I've been searching for all of the things needed to get this conversion running, and interestingly the one thing I can't seem to find is a gcode sender with lathe support that runs on Linux.  IOSender seems to be about the only option but it's Windows only.  I could run Windows on the pi4, but I would really rather not.  Have any ideas here?

C|

**Reply by dresco** [reply]
> As long as all the parts are currently available in the USA.

Passives, connectors etc have come from LCSC, but all the more interesting parts from either Mouser or DigiKey. Only thing that might be an issue is the ILI9488 display with the flex connector, these came from AliExpress.. 

> one thing I can't seem to find is a gcode sender with lathe support that runs on Linux. IOSender seems to be about the only option but it's Windows only. I could run Windows on the pi4, but I would really rather not. Have any ideas here?

There was a discussion (or maybe issue) here, where someone mentioned a porting effort using AvaloniaUI. I feel your pain, haven't found a cross platform sender that I really like (let alone with lathe support).. Don't currently have any Windows machines in the house, so I have to spin up a VM for IOSender testing.

**Comment by cbaugher** [comment]
> 
> Passives, connectors etc have come from LCSC, but all the more interesting parts from either Mouser or DigiKey. Only thing that might be an issue is the ILI9488 display with the flex connector, these came from AliExpress..
> 

I managed to pull a BOM out of KiCAD, and yeah it looks like pretty standard stuff I should be able to get from the usual places.  The display would have to come from China, at least if I only want the bare display.  All of the panels imported to the US seem to be mounted to a carrier board of sorts.  I could probably separate the display from the carrier as long as they aren't glued together.  But then I wouldn't be assured of getting the correct panel.  Hmmm.

If I'm going to have to wait for stuff from China, I might as well just order pre-assembled boards from JLC.  Which means I'm not far from doing mini production run.  What are your thoughts on putting some kits together?  Hehe!


> 
> There was a discussion (or maybe issue) here, where someone mentioned a porting effort using AvaloniaUI. I feel your pain, haven't found a cross platform sender that I really like (let alone with lathe support).. Don't currently have any Windows machines in the house, so I have to spin up a VM for IOSender testing.


I saw that, and I'm tempted to have a go at it.  Problem is I'm not the best programmer, despite actually going to school for that.  LOL!


C|

**Reply by dresco** [reply]
> If I'm going to have to wait for stuff from China, I might as well just order pre-assembled boards from JLC. Which means I'm not far from doing mini production run. What are your thoughts on putting some kits together? Hehe!

Yes sure, leave it with me & I'll work out what JLC need to get some boards assembled. Have only previously requested single sided with stock LCSC parts, so may be a bit of a learning curve..

**Comment by cbaugher** [comment]
> > If I'm going to have to wait for stuff from China, I might as well just order pre-assembled boards from JLC. Which means I'm not far from doing mini production run. What are your thoughts on putting some kits together? Hehe!
> 
> Yes sure, leave it with me & I'll work out what JLC need to get some boards assembled. Have only previously requested single sided with stock LCSC parts, so may be a bit of a learning curve..


Sounds great!  Let me know what you come up with!  It would be fun putting together a run myself, but I'm already way overloaded on projects and really need to stop bringing in more.  <facepalm>

**Comment by fairderben** [comment]
As soon as you come up with a (partial) BOM, Im all in for ordering a PCB and providing tests!

**Comment by mundsen** [comment]
I'm interested 😊

Great project 👏🏻

**Comment by mintracer** [comment]
cool project, keep going, i´m intrested too!

**Comment by kevindt** [comment]
If you are looking for more participants both to help cut the unit costs of prototype board production and for testing, I'm interested.  I have no problem with hardware assembly by hand, but it would be very worthwhile to be able to get a board with the SMT parts in place.  It also sounds like finding a stable source for the displays is a possible problem - though ordering from Ali Express is not an issue.  I'm in the USA, N. California and run grblHAL on Teensy 4.1 on the home CNC.

**Comment by dresco** [comment]
Thanks all for the positive feedback.

I've been working out my actual costs for each of the major components so far, and getting a price for fully assembled main boards. (This wasn't as hard as I'd imagined, main differences to previous boards were that (1) components that aren't already in JLC/LCSC stock need to be pre-ordered into a personal parts library, and (2) double sided assembly naturally adds to the setup costs). Adding the through hole parts for assembly didn't seem to make a huge difference at this point, so including that as well.

If there is sufficient interest, my next steps would be to place an initial order for 10x assembled main boards (along with any other parts I'm missing for ten full kits), and to offer the various parts for sale through an online storefront. 

---

**Note: Am adding my current thoughts on hardware pricing here, in an effort to get a better feel for interest before proceeding. I may come along in the future and update or remove this section...**

Estimated retail pricing is likely to be;

**Front panel -- £15**

**Main board -- £118**
PCB fully assembled on both sides, including all through hole components.
3.5" ILI9488 display.

**Keypad assembly -- £35**
PCB assembled with SMD components.
Though hole components - keys with caps (x32), screw terminals (x6), IDC connector.
Brass standoffs, ribbon cable with IDC connectors.

**Rotary encoder breakouts -- £23 for 3 (feed, spindle, rapid)**
PCBs assembled with SMD components.
Through hole components - Bourns PEC11R rotary encoders, JST PH headers.
Cable assemblies - JST PH connectors with 300mm wire.

My price calculations are higher than I'd been hoping for - especially for the main board. These prices reflect my actual costs for each assembly (typically in 10x quantities), with a commonly recommended cost multiplier to cover time & effort.

I could potentially reduce the multiplier or have a discount code for early adopters / testers, but am left wondering whether the current design is just too expensive to be a viable option at these low quantities?

Would be genuinely interested in any feedback, especially with experience of selling boards or kits in the hobbyist market... 

---

**Reply by cbaugher** [reply]
I would guess most buyers at this point will be experimenters, so unassembled or partially assembled kits will likely be preferred.  Personally I would buy 2 sets of everything except the front panel.  I'm converting a lathe and a mill, so would want to change the panel for the lathe at least.  Probably I would change the mill panel layout a bit as well.

The pricing seems maybe a bit high.  But considering that all the parts are included and the hard assembly is done, it really isn't that bad.  And it saves having to spend time sourcing and ordering components.  I would order these today if they were available.

One minor thing... It would be nice if different colors were available for the button caps.  That way different functions could be color coded.


C|

**Comment by kevindt** [comment]
I don't have experience of selling boards, but I've bought quite a few from low-volume sources over the years.  My feeling is that you might want to define your market(s), since that will have a lot to do with likelihood of sales at different price-points.  I see there being two main markets: 

One being those who see this as a finished product that just works out of the box (and maybe needs a case option).  They're unlikely to play with the firmware or add-ins.  Not sure it's at that stage yet, but the price is not exorbitant for the features offered if it is a clean tested product with ongoing support.  See below for one addition that might increase desirability even for that group.

The others are the experimenters like me.  We are probably more interested in the mainboard and leaving some flexibility in the mechanical design.  That is a fair price assembled and tested.  You might consider offering it a bit cheaper with only the SMT/screen connector assembly done and with/without a kit of the connectors/through-hole parts.  Many others do that (like Phil for his grblHAL boards).  Discounting for early adopters/testers is always welcome, of course.  Getting down to that magic $100 figure by that route and by cutting your assembly time increases salability to that group - who make up a lot of the Tindie buyers, etc.  

I'm willing to buy the whole package to test out, even if I end up not using the keyboard side as-is, but many might not be.  I count the added cost as a contribution to the project - and a significant saving in terms of time on what I was thinking of doing as a one-off.

I've been thinking along somewhat similar lines, before I saw this.
- One good rotary encoder using the built-in functionality on a STM32 timer.  Switchable to different functions via a second simple slow encoder that toggles between jog axis selection and jog distance multiplier selection using its push-switch.
- An RS485 or possibly wireless link to controller.   
- Hardware switches for Cycle Start, Halt.
- Simple slow encoders for feed and spindle over-rides, using the switches on them for feed hold and spindle CW enable/disable.  I never use CCW so no need to bring it out.
After that, my wish-list was a little different.  
- Was not planning on KB jog if I have the good rotary encoder.
- Was not planning on needing A-axis buttons, spindle buttons, encoder jog button group (all on one or another encoder).

I'd thought about hard (hardware) vs, soft (touch) buttons for what I commonly do, some of which are single codes, some are simple small gcode macros definable in the controller:  (You have some of the single codes implemented on buttons).
- Home
- Set workspace zero on each axis
- GTZ on each axis
- Go to safe Z height then go to X0Y0
- Go to safe Z height then rapid to selectable compass points N,NE,E,SE,S,SW,W,NW or center
- Probe at current XY location (perhaps two variants, one at top of conductive surface (PCB), one with input for probe plate height)
- Save current location, go to safe Z height, go to fixed location, wait for tool insertion, got to defined probe location and probe to get initial TLO
- Save current location, go to safe Z height, go to fixed location wait for new tool insertion, got to defined probe location and probe to get tool change TLO

I'm not sure if you're implementing the ability to trigger macros on the screen - haven't looked that deeply into your code yet.  It would be nice to have a row of maybe 10 hardware buttons above the Run/Overrides groups to trigger macros that were defined in the grblHAL controller though.  That avoids having to allow macro programing in the box by end-users.

**Comment by dresco** [comment]
Thanks for the input, some good points there..

> Assembly

I'm thinking a good compromise may be to get just the top layer assembled at the PCB house. That includes all the 'tricky' components & gets me back onto the "economic assembly" option, which helps with the costs. 

The bottom layer is a mixture of SMD and through hole parts, but nothing too difficult. SMD parts are 4x 0603 leds, 2x pushbuttons, USB-C connector, and the optional swd header - all easy enough to hand solder I think. Though hole parts are just standard headers, jumpers, & screw terminals.

> It would be nice if different colors were available for the button caps.

Agreed. I've already ended up using a mixture of white and grey on my panel to differentiate functions.

> I'm not sure if you're implementing the ability to trigger macros on the screen

Not yet, but no reason why not. Would just need to add the extra key definitions & add handlers in the plugin.

> Was not planning on KB jog if I have the good rotary encoder.

Just my personal thoughts on this. I find myself using keyboard jog all the time for rough positioning, and then the rotary encoder for accurately dialling in the last bit.

Advantage of the keyboard jog is that you can tune it to a speed you're comfortable with, and it just stops when you release the key - is quite intuitive for general moving around. (Is implemented to keep adding small jog movements to the queue while pressed, and then a cancel on release).

Rough jogging with the rotary encoder, you just have to be more mindful of not overshooting where you want to end up, so I find it less convenient.
